gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java

```diff
@@ -18,76 +25,296 @@
-      copyJarDependencies (state);
+      suite = CompactionSuiteUtils.getCompactionSuiteFactory(state).createSuite(state);
 
-      DatasetsFinder finder  = DatasetUtils.instantiateDatasetFinder(state.getProperties(),
-               getSourceFileSystem(state),
-               DefaultFileSystemGlobFinder.class.getName());
+      initJobDir(state);
+      copyJarDependencies(state);
+      DatasetsFinder finder = DatasetUtils.instantiateDatasetFinder(state.getProperties(),
+              getSourceFileSystem(state),
+              DefaultFileSystemGlobFinder.class.getName());
 
-      List<CompactionVerifier> verifiers = suite.getDatasetsFinderVerifiers();
       List<Dataset> datasets = finder.findDatasets();
+      CompactionWorkUnitIterator workUnitIterator = new CompactionWorkUnitIterator ();
+
+      // Spawn a single thread to create work units
+      new Thread(new SingleWorkUnitGeneratorService (state, datasets, workUnitIterator)).start();
+      return new BasicWorkUnitStream.Builder (workUnitIterator).build();
+
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A work unit generator service will do the following:
+   * 1) Convert dataset iterator to verification callable iterator, each callable element is a verification procedure
+   * 2) Use {@link IteratorExecutor} to execute callable iterator
+   * 3) Collect all failed datasets at step 2), retry them until timeout. Once timeout create failed tasks on purpose.
+   */
+  private class SingleWorkUnitGeneratorService implements Runnable {
+    private SourceState state;
+    private List<Dataset> datasets;
+    private CompactionWorkUnitIterator workUnitIterator;
+    private IteratorExecutor executor;
+
+    public SingleWorkUnitGeneratorService (SourceState state, List<Dataset> datasets, CompactionWorkUnitIterator workUnitIterator) {
+      this.state = state;
+      this.datasets = datasets;
+      this.workUnitIterator = workUnitIterator;
+    }
+
+    public void run () {
+      try {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        int parallel = this.state.getPropAsInt(CompactionVerifier.VERIFICATION_PARALLEL, 5);
+        long timeOut = this.state.getPropAsInt(CompactionVerifier.VERIFICATION_TIMEOUT, 60);
+        while (datasets.size() > 0) {
+          Iterator<Callable<VerifiedDataset>> verifierIterator =
+                  Iterators.transform (datasets.iterator(), new Function<Dataset, Callable<VerifiedDataset>>() {
+                    @Nullable
+                    @Override
+                    public Callable<VerifiedDataset> apply(Dataset dataset) {
+                      return new DatasetVerifier (dataset, workUnitIterator, suite.getDatasetsFinderVerifiers());
+                    }
+                  });
+
+          executor = new IteratorExecutor<>(verifierIterator, parallel,
+                  ExecutorsUtils.newDaemonThreadFactory(Optional.of(log), Optional.of("Verifier-compaction-dataset-pool-%d")));
```

Comment:
This should not be a daemon factory, verifications in progress should prevent the jvm from exiting.